### PR TITLE
add incoming / outgoing payment hash commands

### PR DESCRIPTION
**Pull request by matthappens**
Ref: https://linear.app/lightsparkdev/issue/LIG-6130/add-invoiceforpaymenthash-outgoingpaymentsforpaymenthash-and
Adding IncomingPaymentsForPaymentsHash, OutgoingPaymentsForPaymentsHash, and InvoiceForPaymentsHash

### CODE REVIEW

**Comment by matthappens** [comment]
* **#197** <a href="https://app.graphite.dev/github/pr/lightsparkdev/kotlin-sdk/197?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `develop`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @matthappens and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: lightspark-sdk/src/commonMain/kotlin/com/lightspark/sdk/LightsparkSyncClient.kt [hunk 1]
```diff
@@ -563,7 +563,7 @@ class LightsparkSyncClient constructor(config: ClientConfig) {
     /**
      * Fetch incoming payments for a given payment hash.
      *
-     * @param paymentHash The payment hash of the invoice for which to fetch the incoming payments.
```
Comment (zhenlu) [inline lightspark-sdk/src/commonMain/kotlin/com/lightspark/sdk/LightsparkSyncClient.kt:566]: unintended change?

**Review by zhenlu — APPROVED** [review]

File: lightspark-sdk/src/commonMain/kotlin/com/lightspark/sdk/LightsparkCoroutinesClient.kt [hunk 2]
```diff
@@ -1036,6 +1036,90 @@ class LightsparkCoroutinesClient private constructor(
...
+                    requireNotNull(it["outgoing_payments_for_payment_hash"]) { "No payment output found in response" }
+                val paymentsJson =
+                    requireNotNull(outputJson.jsonObject["payments"]) { "No payments found in response" }
+                serializerFormat.decodeFromJsonElement(paymentsJson)
+            },
+        )
+    }
+
+    /**
+     * fetch invoice for a given payments hash
```
Comment (jklein24) [inline lightspark-sdk/src/commonMain/kotlin/com/lightspark/sdk/LightsparkCoroutinesClient.kt:1070]:
```suggestion
     * Fetch invoice for a given payment hash
```

File: lightspark-sdk/src/commonMain/kotlin/com/lightspark/sdk/LightsparkSyncClient.kt [hunk 3]
```diff
@@ -576,6 +576,28 @@ class LightsparkSyncClient constructor(config: ClientConfig) {
...
     }
 
+    /**
+     * fetch outgoing payments for a given payment hash
+     * 
+     * @param paymentHash the payment hash of the invoice for which to fetch the outgoing payments
+     * @param transactionStatuses the transaction statuses to filter the payments by.  If null, all payments will be returned.
+     */
+    @Throws(LightsparkException::class, LightsparkAuthenticationException::class, CancellationException::class)
+    fun getOutgoingPaymentsForPaymentsHash(
```
Comment (jklein24) [inline lightspark-sdk/src/commonMain/kotlin/com/lightspark/sdk/LightsparkSyncClient.kt:586]: Do we also need InvoiceForPaymentHash in this file?

File: lightspark-sdk/src/commonMain/kotlin/com/lightspark/sdk/graphql/IncomingPaymentsForInvoice.kt [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+import com.lightspark.sdk.model.IncomingPayment
+
+const val IncomingPaymentsForInvoiceQuery = """
+query IncomingPaymentsForInvoice(
+    ${'$'}invoiceId: Hash32!,
+    ${'$'}transactionStatuses: [TransactionStatus!] = null
+) {
+    incoming_payments_for_invoice_query(input: {
+        invoice_id: ${'$'}invoiceId,
+         statuses: ${'$'}transactionStatuses
```
Comment (jklein24) [inline lightspark-sdk/src/commonMain/kotlin/com/lightspark/sdk/graphql/IncomingPaymentsForInvoice.kt:12]: Nit:
```suggestion
        statuses: ${'$'}transactionStatuses
```

File: lightspark-sdk/src/commonTest/kotlin/com/lightspark/sdk/ClientIntegrationTests.kt [hunk 5]
```diff
@@ -309,6 +309,46 @@ class ClientIntegrationTests {
         payments[0].id.shouldBe(payment?.id)
     }
 
+    // todo add payment has tests
```
Comment (jklein24) [inline lightspark-sdk/src/commonTest/kotlin/com/lightspark/sdk/ClientIntegrationTests.kt:312]: What do you mean by this one? Did you mean to remove this todo?

File: lightspark-sdk/src/commonTest/kotlin/com/lightspark/sdk/ClientIntegrationTests.kt [hunk 6]
```diff
@@ -309,6 +309,46 @@ class ClientIntegrationTests {
...
+        var outgoingPayment: OutgoingPayment? = client.payInvoice(node.id, invoice, maxFeesMsats = 100_000)
+        outgoingPayment.shouldNotBeNull()
+        while (outgoingPayment?.status == TransactionStatus.PENDING) {
+            delay(500)
+            outgoingPayment = OutgoingPayment.getOutgoingPaymentQuery(outgoingPayment.id).execute(client)
+            println("Payment status: ${outgoingPayment?.status}")
+        }
+        outgoingPayment?.status.shouldBe(TransactionStatus.SUCCESS)
+
+        val payments = client.getOutgoingPaymentsForPaymentsHash(outgoingPayment?.transactionHash!!)
```
Comment (jklein24) [inline lightspark-sdk/src/commonTest/kotlin/com/lightspark/sdk/ClientIntegrationTests.kt:328]: Does this work? I don't think the transactionHash is the right hash here. It should actually be the invoice's payment hash.

File: lightspark-sdk/src/commonTest/kotlin/com/lightspark/sdk/ClientIntegrationTests.kt [hunk 7]
```diff
@@ -309,6 +309,46 @@ class ClientIntegrationTests {
...
+        val testInvoice = client.createInvoice(node.id, 100_000, "test invoice")
+        var payment: IncomingPayment? = client.createTestModePayment(node.id, testInvoice.data.encodedPaymentRequest)
+        payment.shouldNotBeNull()
+        while (payment?.status == TransactionStatus.PENDING) {
+            delay(500)
+            payment = IncomingPayment.getIncomingPaymentQuery(payment.id).execute(client)
+            println("Payment status: ${payment?.status}")
+        }
+
+        val invoice = client.getInvoiceForPaymentHash(payment?.transactionHash!!)
```
Comment (jklein24) [inline lightspark-sdk/src/commonTest/kotlin/com/lightspark/sdk/ClientIntegrationTests.kt:347]: Same note here.

**Review by jklein24 — CHANGES_REQUESTED** [review]

File: lightspark-sdk/src/commonTest/kotlin/com/lightspark/sdk/ClientIntegrationTests.kt (hunk 6 above)
Comment (zhenlu) [reply lightspark-sdk/src/commonTest/kotlin/com/lightspark/sdk/ClientIntegrationTests.kt:328]: They should be the same hash iiuc

Comment (matthappens) [reply lightspark-sdk/src/commonTest/kotlin/com/lightspark/sdk/ClientIntegrationTests.kt:328]: that being said, even if it's the same hash, for clarity it should probably come from the invoice payment hash?

File: lightspark-sdk/src/commonTest/kotlin/com/lightspark/sdk/ClientIntegrationTests.kt [hunk 8]
```diff
@@ -309,6 +309,44 @@ class ClientIntegrationTests {
...
+        var outgoingPayment: OutgoingPayment? = client.payInvoice(node.id, invoice, maxFeesMsats = 100_000)
+        outgoingPayment.shouldNotBeNull()
+        while (outgoingPayment?.status == TransactionStatus.PENDING) {
+            delay(500)
+            outgoingPayment = OutgoingPayment.getOutgoingPaymentQuery(outgoingPayment.id).execute(client)
+            println("Payment status: ${outgoingPayment?.status}")
+        }
+        outgoingPayment?.status.shouldBe(TransactionStatus.SUCCESS)
+
+        val payments = client.getOutgoingPaymentForPaymentHash(outgoingPayment?.transactionHash!!)
```
Comment (matthappens) [inline lightspark-sdk/src/commonTest/kotlin/com/lightspark/sdk/ClientIntegrationTests.kt:326]: I still need to fix this one.  the invoice created on line 316 is just a string, is there a way to infer the payment has from that?

File: lightspark-sdk/src/commonTest/kotlin/com/lightspark/sdk/ClientIntegrationTests.kt (hunk 5 above)
Comment (matthappens) [reply lightspark-sdk/src/commonTest/kotlin/com/lightspark/sdk/ClientIntegrationTests.kt]: ah this was junk code from trying out graphite, removing

File: lightspark-sdk/src/commonTest/kotlin/com/lightspark/sdk/ClientIntegrationTests.kt (hunk 8 above)
Comment (zhenlu) [reply lightspark-sdk/src/commonTest/kotlin/com/lightspark/sdk/ClientIntegrationTests.kt:326]: One way to do this is to call decodeInvoice for that invoice, then you'll get the InvoiceData which contains paymentHash

https://github.com/lightsparkdev/kotlin-sdk/blob/6342b0fea7379b69fbb4141ea760cff92604d695/wallet-sdk/src/commonMain/kotlin/com/lightspark/sdk/wallet/model/InvoiceData.kt#L27

Comment (zhenlu) [reply lightspark-sdk/src/commonTest/kotlin/com/lightspark/sdk/ClientIntegrationTests.kt:326]: Another way to do this without network request is to bech32 decode the invoice, and find the p field in it.

https://github.com/lightning/bolts/blob/master/11-payment-encoding.md#tagged-fields

**Review by jklein24 — APPROVED** [review]

**Pull request merged**